tests-cpp/test-layout-builder.cpp

```diff
@@ -0,0 +1,1000 @@
+
+  auto form = builder.form();
+
+  assert (form ==
+  "{ "
+      "\"class\": \"NumpyArray\", "
+      "\"primitive\": \"char\", "
+      "\"form_key\": \"node0\" "
+  "}");
+
+  char* ptr0 = new char[builder.length()];
+  builder.to_buffers(ptr0);
+
+  dump("node0", ptr0, builder.length());
+  std::cout<<std::endl;
+}
+
+void
+test_numpy_double() {
+  auto builder = awkward::NumpyLayoutBuilder<0, initial, double>();
+
+  size_t data_size = 9;
+
+  double data[9] = {1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9};
+
+  builder.append(data, data_size);
+
+  // [1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9]
+
+  auto form = builder.form();
+
+  assert (form ==
+  "{ "
+      "\"class\": \"NumpyArray\", "
+      "\"primitive\": \"float64\", "
+      "\"form_key\": \"node0\" "
+  "}");
+
+  double* ptr0 = new double[builder.length()];
+  builder.to_buffers(ptr0);
+
+  dump("node0", ptr0, builder.length());
+  std::cout<<std::endl;
+}
+
+void
+test_numpy_complex() {
+  auto builder = awkward::NumpyLayoutBuilder<0, initial, std::complex<double>>();
+
+  builder.append({1.1, 0.1});
+  builder.append({2.2, 0.2});
+  builder.append({3.3, 0.3});
+  builder.append({4.4, 0.4});
+  builder.append({5.5, 0.5});
+
+  // [1.1 + 0.1j, 2.2 + 0.2j, 3.3 + 0.3j, 4.4 + 0.4j, 5.5 + 0.5j]
+
+  auto form = builder.form();
+
+  assert (form ==
+  "{ "
+      "\"class\": \"NumpyArray\", "
+      "\"primitive\": \"complex128\", "
+      "\"form_key\": \"node0\" "
+  "}");
+
+  std::complex<double>* ptr0 = new std::complex<double>[builder.length()];
+  builder.to_buffers(ptr0);
+
+  dump("node0", ptr0, builder.length());
+  std::cout<<std::endl;
+}
+
+void
+test_list_offset_of_numpy() {
+  auto builder = awkward::ListOffsetLayoutBuilder<0,
+      initial, awkward::NumpyLayoutBuilder<1, initial, double>
+  >();
+
+  auto builder2 = builder.begin_list();
+  builder.append(1.1);
+  builder.append(2.2);
+  builder.append(3.3);
+  builder.end_list();
+
+  builder.begin_list();
+  builder.end_list();
+
+  builder.begin_list();
+  builder.append(4.4);
+  builder.append(5.5);
+  builder.end_list();
+
+  // [[1.1, 2.2, 3.3], [], [4.4, 5.5]]
+
+  auto form = builder.form();
+
+  assert (form ==
+  "{ "
+      "\"class\": \"ListOffsetArray\", "
+      "\"offsets\": \"i64\", "
+      "\"content\": { "
+          "\"class\": \"NumpyArray\", "
+          "\"primitive\": \"float64\", "
+          "\"form_key\": \"node1\" "
+      "}, "
+      "\"form_key\": \"node0\" "
+  "}");
+
+  int64_t* ptr0 = new int64_t[builder.length() + 1];
+  builder.to_buffers(ptr0);
+
+  double* ptr1 = new double[builder2->length()];
+  builder2->to_buffers(ptr1);
+
+  dump("node0", ptr0, builder.length() + 1,
+       "node1", ptr1, builder2->length());
+  std::cout<<std::endl;
+}
+
+void
+test_list_offset_of_list_offset() {
+  auto builder = awkward::ListOffsetLayoutBuilder<0,
+      initial, awkward::ListOffsetLayoutBuilder<1,
+          initial, awkward::NumpyLayoutBuilder<2, initial, double>
+  >>();
```

Comment:
Having to manually set ids is pretty user-unfriendly. They only have to be unique, and they don't have to be created quickly. Couldn't there be a global

```c++
static std::atomic<uint64_t> layoutbuilder_next_id = 0;
```

with helper functions to give every new LayoutBuilder instance a new id? They only have to be unique per-tree and that would make them unique per-process, but being _too_ unique isn't a problem.

Alternatively, each LayoutBuilder constructor could initiate a recursive walk that reassigns ids in-place to be unique per-tree (with the added benefit of being repeatable per-tree). It's okay if the act of inserting a LayoutBuilder into a tree changes its id because LayoutBuilders are mutable objects that can't be in more than one tree, anyway. (On second thought, this recursive-replace method is sounding better than the global atomic integer, since building the same tree would produce the same Form, and that's a reproducibility benefit.)